File: reto_clean_code/angular/step1-oop.ts
// Refactorización orientada a objetos de la transferencia bancaria.  Se
// introducen clases para encapsular el estado de las cuentas y la lógica
// de la operación.

export class CuentaBancaria {
  constructor(public titular: string, public saldo: number) {}

  retirar(monto: number): boolean {
    if (monto <= this.saldo) {
      this.saldo -= monto;
      return true;
    }
    return false;
  }

  depositar(monto: number): void {
    this.saldo += monto;
  }
}

export class Transferencia {
  realizar(origen: CuentaBancaria, destino: CuentaBancaria, monto: number): void {
    if (origen.retirar(monto)) {
      destino.depositar(monto);
      console.log(`Transferencia de ${monto} de ${origen.titular} a ${destino.titular}`);
    } else {
      console.log('Saldo insuficiente.');
    }
  }
}

File: reto_clean_code/python/step1_oop.py
"""
Ejemplo orientado a objetos para realizar una transferencia bancaria.
Las clases encapsulan el estado de las cuentas y la lógica de la operación.
"""


class CuentaBancaria:
    def __init__(self, titular: str, saldo: float) -> None:
        self.titular = titular
        self.saldo = saldo

    def retirar(self, monto: float) -> bool:
        if monto <= self.saldo:
            self.saldo -= monto
            return True
        return False

    def depositar(self, monto: float) -> None:
        self.saldo += monto


class Transferencia:
    def realizar(self, origen: 'CuentaBancaria', destino: 'CuentaBancaria', monto: float) -> None:
        if origen.retirar(monto):
            destino.depositar(monto)
            print(f"Transferencia de {monto} de {origen.titular} a {destino.titular}")
        else:
            print("Saldo insuficiente")


if __name__ == "__main__":
    cuenta_a = CuentaBancaria('A', 1000)
    cuenta_b = CuentaBancaria('B', 500)
    transferencia = Transferencia()
    transferencia.realizar(cuenta_a, cuenta_b, 200)
    print(cuenta_a.saldo, cuenta_b.saldo)